### PR TITLE
use checked arithmetic

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem
Should use checked arithmetic in most cases.

#### Summary of Changes
- Add tests
- Use checked arithmetic
- Add feature gate

Feature Gate Issue: #31273

### CODE REVIEW

File: runtime/src/accounts.rs [hunk 1]
```diff
@@ -594,7 +594,10 @@ impl Accounts {
             }
         };
 
-        if payer_account.lamports() < fee + min_balance {
+        if payer_account.lamports() < fee.checked_add(min_balance).ok_or_else(|| {
```
Comment (taozhu-chicago) [inline runtime/src/accounts.rs:597]: Saturating math probably not work here:  `u64::MAX < u64::MAX.saturating_add(1)` should be true but not; similar `1.saturating_sub(99)<0` should be true but also not.

Using `checked_add` instead, cause if `fee.checked_add(min_balance)` exceeds boundary, the payer surely won't be able to pay, even it has u64::MAX balance.

File: runtime/src/accounts.rs [hunk 2]
```diff
@@ -594,7 +594,12 @@ impl Accounts {
...
-        if payer_account.lamports() < fee + min_balance {
+        let required_lamports = fee.checked_add(min_balance).ok_or_else(|| {
+            error_counters.insufficient_funds += 1;
+            TransactionError::InsufficientFundsForFee
+        })?;
+
+        if payer_account.lamports() < required_lamports {
             error_counters.insufficient_funds += 1;
             return Err(TransactionError::InsufficientFundsForFee);
         }
```
Comment (t-nelson) [inline runtime/src/accounts.rs:605]:
```suggestion
        payer_account
            .lamports()
            .checked_sub(min_balance)
            .and_then(|v| v.checked_sub(fee))
            .ok_or_else(|| {
                error_counters.insufficient_funds += 1;
                TransactionError::InsufficientFundsForFee
            })?;
```

maybe combine them?

File: runtime/src/accounts.rs (hunk 1 above)
Comment (t-nelson) [reply runtime/src/accounts.rs]: yeah nice catch!

File: runtime/src/accounts.rs (hunk 2 above)
Comment (taozhu-chicago) [reply runtime/src/accounts.rs]: Neat!

File: runtime/src/accounts.rs [hunk 3]
```diff
@@ -4319,4 +4329,196 @@ mod tests {
             (Err(TransactionError::InsufficientFundsForFee), None),
         );
     }
+
+    struct TestParameter {
```
Comment (t-nelson) [inline runtime/src/accounts.rs:4333]:
```suggestion
    struct ValidateFeePayerTestParameter {
```

just a nit

**Review by t-nelson — COMMENTED** [review]
i think i'm down to just nits

File: runtime/src/accounts.rs [hunk 4]
```diff
@@ -594,10 +594,20 @@ impl Accounts {
...
+                .lamports()
+                .checked_sub(min_balance)
+                .and_then(|v| v.checked_sub(fee))
+                .ok_or_else(|| {
+                    error_counters.insufficient_funds += 1;
+                    TransactionError::InsufficientFundsForFee
+                })?;
+        } else if payer_account.lamports() < fee + min_balance {
             error_counters.insufficient_funds += 1;
             return Err(TransactionError::InsufficientFundsForFee);
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs:608]: You are right on it. It was my original intention too. Think `fmt` was cool with it, but `clippy` wasn't. Let me double check, can allow clippy to help removing gate safer.

[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/accounts.rs]: 41acd270b4b1cd509b0a79ae9e5c39f367f17493

File: runtime/src/accounts.rs [hunk 5]
```diff
@@ -594,10 +594,24 @@ impl Accounts {
             }
         };
 
-        if payer_account.lamports() < fee + min_balance {
-            error_counters.insufficient_funds += 1;
-            return Err(TransactionError::InsufficientFundsForFee);
+        // allow collapsible-else-if to help removing gate safer.
```
Comment (mvines) [inline runtime/src/accounts.rs:597]:
```suggestion
        // allow collapsible-else-if to make removing the feature gate safer once activated
```

**Review by t-nelson — APPROVED** [review]
:shipit:

**Pull request merged**